spec/index.bs

```diff
@@ -873,6 +927,103 @@ interface GPUBufferUsage {
+            1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapping pending for reading=].
+            1. Enqueue an operation on the [=Queue timeline=] that will execute the following:
+
+                1. Let |m| be a new {{ArrayBuffer}} of size the {{[[size]]}} of |this|.
+                1. Set the content of |m| to the content of |this|'s allocation.
+                1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapped for reading=].
+                1. If |p| is pending:
+
+                    1. Resolve |p| with |m|.
+
+            1. Return |p|.
+        </div>
+</dl>
+
+### {{GPUBuffer/mapWriteAsync|GPUDevice.mapWriteAsync}} ### {#GPUBuffer-mapWriteAsync}
+
+<dl dfn-type="method" dfn-for="GPUBuffer">
+    : <dfn>mapWriteAsync()</dfn>
+    ::
+        <div algorithm="GPUBuffer.mapWriteAsync()">
+            <!-- TODO(kangz) handle error buffers once we have a description of the error monad -->
+            1. If the {{[[usage]]}} slot of |this| doesn't contain the {{GPUBufferUsage/MAP_WRITE}} bit:
+
+                1. Record a validation error on the current scope.
+                1. Return a new promise that rejects. <!-- TODO(kangz) specify on that it happens on the device timeline-->
+
+            1. Let |p| be a new {{Promise}}.
+            1. Set the {{[[mapping_promise]]}} slot of |this| to |p|.
+            1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapping pending for writing=].
+            1. Enqueue an operation on the [=Queue timeline=] that will execute the following:
+
+                1. Let |m| be a new {{ArrayBuffer}} of size the {{[[size]]}} of |this| that is filled with zeroes.
+                1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapped for writing=].
+                1. If |p| is pending:
+
+                    1. Resolve |p| with |m|.
+
+            1. Return |p|.
+        </div>
+</dl>
+
+### {{GPUBuffer/unmap|GPUBuffer.unmap()}} ### {#GPUBuffer-unmap}
+
+<dl dfn-type="method" dfn-for="GPUBuffer">
+    : <dfn>unmap()</dfn>
+    ::
+        <div algorithm="GPUBuffer.unmap()">
+
+            1. If the {{[[state]]}} slot of |this| is [=buffer state/unmapped=] or [=buffer state/destroyed=]:
+
+                1. Record a validation error on the current scope.
+                1. Return.
+
+            1. If the {{[[mapping_promise]]}} slot of |this| is not null:
+
+                1. Reject {{[[mapping_promise]]}}.
+                1. Set the {{[[mapping_promise]]}} slot of |this| to null.
+
+            1. If the {{[[mapping]]}} slot of |this| is not null:
+
+                1. If the {{[[state]]}} slot of this is [=buffer state/mapped for writing=]:
+
+                    1. Enqueue an operation on the [=Queue timeline=] that updates |this|'s allocation to the content of the {{ArrayBuffer}} in the {{[[mapping]]}} slot of |this|.
+
+                1. Detach the {{ArrayBuffer}} in the {{[[mapping]]}} slot of |this|.
```

Comment:
Should we be more specific and say the array buffer object is detached from its contents?

Reply:
I tried linking to http://www.ecma-international.org/ecma-262/6.0/#sec-detacharraybuffer bit couldn't figure out how. Reworded to make more explicit.